### ISSUE TITLE
Support for different tensorization strategies

### ISSUE THREAD

**Issue by Oufattole** [mmcdermott/MEDS_transforms#82]
The tensorization stage currently uses the function [convert_to_NRT](https://github.com/mmcdermott/MEDS_transforms/blob/afbcdac458394e01251e05182a7c454a1fe6203e/src/MEDS_transforms/transforms/tensorization.py#L19) to transform data from parquet format to a nested ragged tensor, which facilitates the fast loading of individual patient data during training sessions. However, there is also a need to support other versions of this function due to the significant post-processing involved in adapting the data for different model formats. For example the ETHOS format requires quantiling and the prompt expanded triplet format (e.g., [TS] [TS = 11:13pm] [CODE: HR] [VAL] [VAL = 88] [CODE: DX] [CODE: BP/SYS] [VAL] [VAL = 120] ...), further detailed in [this meds-torch issue](https://github.com/Oufattole/meds-torch/issues/30) requires generating up to three events for each observation (essentially requiring a for loop through every observation in the dataset).

To start off, added an inefficient implementaiton of the prompt expanded triplet tensorization, which could be optionally supported by introducing a tensorization stage argument that allows the user to select the desired tensorization method, which will dynamically choose the corresponding `convert_to_NRT` function to use. I added the [convert_to_prompt_expanded_observation_NRT](https://github.com/mmcdermott/MEDS_transforms/blob/df4ded6f05e29ac75c0355655946b30d766d1264/src/MEDS_transforms/transforms/tensorization.py#L24),  and maybe others can be added to the tensorization.py file to accommodate this flexibility.

What do you think @mmcdermott @prenc @aleksiakolo?

**Comment by mmcdermott** [comment]
I think it might be easier to do some of this earlier in the model
pre-processing pipeline than the tensorization stage. E.g., we can add time
interval tokens in one transform; add value bin tokens in another, and end
up with something tjat can be tensorized in much the same way as other
data. The big difference I see in tensorization modes is whether we are
tensorizing the data in a 2D (flat sequence) or 3D (nested within
timestamp) mode, and whether we are tensorizing time deltas at all. Would
that be a viable solution, do you think?

**Comment by Oufattole** [comment]
Yeah the fact that I had to add [manual code offsets](https://github.com/mmcdermott/MEDS_transforms/blob/df4ded6f05e29ac75c0355655946b30d766d1264/src/MEDS_transforms/transforms/tensorization.py#L19) I think indicates this is a bit late in the pipeline. So maybe a `token_modifier` transform should be added (I'm not attached to that name though). I'm not sure where is best in the pipeline (i.e.e what stage should be the expected input). 

For tensorizing in 2D (flat sequence) mode, incorporating timestamp offsets in the 2D sequence is probably a good idea, specifically for tasks like supervised training or contrastive learning, where in the pytorch_dataset we have to align the dynamic data with their corresponding labels or windows.

**Comment by mmcdermott** [comment]
I would suggest we create a few issues:
  1. A transform to add time interval tokens before unique timestamps.
  2. A transform to separate code tokens from their numerical values.
  3. An issue to update the tensorization code to allow for 2D or 3D tensorization and to permit including or excluding some aspects (e.g., time deltas). 
 
Then, with these issues, you can just add to the data pipeline the add time intervals and separate codes from values transforms before normalizing numerical values and fitting the code vocabulary and update the parameters for the tensorization transform. Would that work @prenc and @Oufattole?

**Comment by mmcdermott** [comment]
Worth noting that issue 1 in the above list can likely be added as simply another time-derived measurement alongside those already supported in this transform: https://github.com/mmcdermott/MEDS_transforms/blob/main/src/MEDS_transforms/transforms/add_time_derived_measurements.py

**Comment by mmcdermott** [comment]
Issue 1 (time interval tokens) is now tracked here: #120 
Issue 2 (separating codes and numeric values), I think, can actually potentially be covered by either #58 or, ideally a combination of match and revise (#101) and extracting values (#121). See the comment on #58 for how to do that.
Issue 3 is now tracked here: https://github.com/mmcdermott/MEDS_transforms/issues/122

Barring further comments, I'm going to close this issue in favor of those linked above.
